runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -6937,16 +6937,69 @@ J9::Z::TreeEvaluator::VMifInstanceOfEvaluator(TR::Node * node, TR::CodeGenerator
    return NULL;
    }
 
+/**
+    * Generates a quick runtime test for valueType node and in case if node is of valueType, generates a branch to helper call
+    * 
+    * @param node monent or monexit node
+    * @param mergeLabel TR::LabelSymbol* to pointing to mergePoint in mainline
+    * @param callLabel TR::LabelSymbol* pointing to the helper Call Label
+    * @param cg CodeGenerator Object Ptr
+    * @return TR::Register* If Given node required a runtime test to check for valueType, it will return the register containing J9Class of given object
+    */
+static TR::Register*
+generateCheckForValueTypeMonitorEnterOrExit(TR::Node *node, TR::LabelSymbol* mergeLabel, TR::LabelSymbol *helperCallLabel, TR::CodeGenerator *cg)
+   {
+   TR::Register *objReg = cg->evaluate(node->getFirstChild());
+   TR::Register *objectClassReg = cg->allocateRegister();
+   TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, objectClassReg, generateS390MemoryReference(objReg, TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);
+   TR::Register *classFlagsRegister = cg->allocateRegister();
+   TR_J9VMBase *fej9 = static_cast<TR_J9VMBase *>(cg->comp()->fe());
+   generateRXInstruction(cg, TR::InstOpCode::L, node, classFlagsRegister, generateS390MemoryReference(objectClassReg, static_cast<uint32_t>(fej9->getOffsetOfClassFlags()), cg));
+   bool generateOOLSection = helperCallLabel == NULL;
+   if (generateOOLSection)
+      helperCallLabel = generateLabelSymbol(cg);
+   
+   generateRIInstruction(cg, TR::InstOpCode::TMLL, node, classFlagsRegister, J9ClassIsValueType);
```

Comment:
This is dangerous. We are assuming the value of `J9ClassIsValueType` is in a specific bit range. If that were ever to change we would have a silent bug here. Can we add a fatal assert that `0 < J9ClassIsValueType <= 0xFFFF`?

Reply:
I have put the static assert same as x86 here (https://github.com/eclipse/openj9/compare/2a2c5fa991bf7b3579cbe91338977bcb1f0e2884..247123bcd0bf7c87262e954972690bf87f08e56a)